exec/java-exec/src/main/java/org/apache/drill/exec/work/metadata/MetadataProvider.java

```diff
@@ -0,0 +1,451 @@
+import org.apache.drill.exec.store.SchemaTreeProvider;
+import org.apache.drill.exec.store.ischema.InfoSchemaConstants;
+import org.apache.drill.exec.store.ischema.InfoSchemaFilter;
+import org.apache.drill.exec.store.ischema.InfoSchemaFilter.ConstantExprNode;
+import org.apache.drill.exec.store.ischema.InfoSchemaFilter.ExprNode;
+import org.apache.drill.exec.store.ischema.InfoSchemaFilter.FieldExprNode;
+import org.apache.drill.exec.store.ischema.InfoSchemaFilter.FunctionExprNode;
+import org.apache.drill.exec.store.ischema.InfoSchemaTableType;
+import org.apache.drill.exec.store.ischema.Records.Catalog;
+import org.apache.drill.exec.store.ischema.Records.Column;
+import org.apache.drill.exec.store.ischema.Records.Schema;
+import org.apache.drill.exec.store.ischema.Records.Table;
+import org.apache.drill.exec.store.pojo.PojoRecordReader;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Contains worker {@link Runnable} classes for providing the metadata and related helper methods.
+ */
+public class MetadataProvider {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(MetadataProvider.class);
+
+  private static final String LIKE_FUNCTION = "like";
+  private static final String AND_FUNCTION = "booleanand";
+  private static final String OR_FUNCTION = "booleanor";
+
+  /**
+   * Super class for all metadata provider runnable classes.
+   */
+  public abstract static class MetadataRunnable implements Runnable {
+    protected final UserClientConnection connection;
+    protected final DrillbitContext dContext;
+    protected final ResponseSender responseSender;
+
+    protected MetadataRunnable(final UserClientConnection connection, final DrillbitContext dContext,
+        final ResponseSender responseSender) {
+      this.connection = Preconditions.checkNotNull(connection);
+      this.dContext = Preconditions.checkNotNull(dContext);
+      this.responseSender = Preconditions.checkNotNull(responseSender);
+    }
+  }
+
+  /**
+   * Runnable that fetches the catalog metadata for given {@link GetCatalogsReq} and sends response at the end.
+   */
+  public static class CatalogsProvider extends MetadataRunnable {
+    private final GetCatalogsReq req;
+
+    public CatalogsProvider(final UserClientConnection connection, final DrillbitContext dContext,
+        final GetCatalogsReq req, final ResponseSender responseSender) {
+      super(connection, dContext, responseSender);
+      this.req = Preconditions.checkNotNull(req);
+    }
+
+    @Override
+    public void run() {
+      final GetCatalogsResp.Builder respBuilder = GetCatalogsResp.newBuilder();
+
+      final InfoSchemaFilter filter = createInfoSchemaFilter(
+          req.hasCatalogNameFilter() ? req.getCatalogNameFilter() : null, null, null, null);
+
+      try (SchemaTreeProvider schemaTreeProvider = new SchemaTreeProvider(dContext)) {
```

Comment:
Why do we create a SchemaTreeProvider per call? Can one instance live in Drillbit or UserSession?